pkg/agent/downloader.go

```diff
@@ -0,0 +1,85 @@
+package agent
+
+import (
+	"fmt"
+	pro "github.com/kubeflow/kfserving/pkg/agent/protocols"
+	"hash/fnv"
+	"log"
+	"os"
+	"path/filepath"
+	"regexp"
+	"strings"
+)
+
+type Downloader struct {
+	// TODO: Add back-off retries
+	ModelDir   string
+	ProtocolManagers map[pro.Protocol]pro.ProtocolManager
+}
+
+var SupportedProtocols = []pro.Protocol{pro.S3}
+
+func (d *Downloader) DownloadModel(event EventWrapper) error {
+	modelSpec := event.ModelSpec
+	if modelSpec != nil {
+		modelUri := modelSpec.StorageURI
+		hashString := hash(modelUri)
+		log.Println("Processing:", modelUri, "=", hashString)
+		successFile := filepath.Join(d.ModelDir, fmt.Sprintf("SUCCESS.%d", hashString))
+		if !pro.FileExists(successFile) {
+			downloadErr := d.download(modelUri)
+			if downloadErr != nil {
+				return fmt.Errorf("download error: %v", downloadErr)
+			} else {
+				file, createErr := os.Create(successFile)
+				if createErr != nil {
+					return fmt.Errorf("create file error: %v", createErr)
+				}
+				defer file.Close()
+			}
+		} else {
+			log.Println("Model", modelSpec.StorageURI, "exists already")
```

Comment:
We should use a logger like zap.S().Infof() or something similar to what we have in the controller.

Reply:
I had issues with getting it to print in the controller, but we can do that last. I will punt on this, _this_ iteration